ocaml/xapi/xapi_pool_update.ml

```diff
@@ -344,11 +344,17 @@ let introduce ~__context ~vdi =
   try
     let update = Db.Pool_update.get_by_uuid ~__context ~uuid:update_info.uuid in
     let vdi_of_update = Db.Pool_update.get_vdi ~__context ~self:update in
-    if vdi <> vdi_of_update then begin
-      Db.Pool_update.set_vdi ~__context ~self:update ~value:vdi
-    end;
-    update
-  with _ ->
+    if vdi_of_update = Ref.null then begin
+        Db.Pool_update.set_vdi ~__context ~self:update ~value:vdi;
+        update
+    end 
+    else if vdi <> vdi_of_update then 
+        raise (Api_errors.Server_error(Api_errors.update_already_exists, [update_info.uuid]))
+    else
+        update
+  with
+  | Api_errors.Server_error(code, _) as e when code = Api_errors.update_already_exists -> raise e
+  | _ ->
```

Comment:
I see. Rather than checking whether it's null or not, we should probably check whether it's valid or not (Db.is_valid_ref) - that'll handle VDI.destroy too.